rfcs/004 - Better Policy UX RFC.md

```diff
@@ -0,0 +1,156 @@
+# 004 - Better Policy UX RFC
+
+## Current Status
+
+### Proposed
+
+2022-09-06
+
+### Accepted
+
+YYYY-MM-DD
+
+#### Approvers
+
+- Full Name <email@example.com>
+
+### Implementation
+
+- [ ] Implemented: [One or more PRs](https://github.com/iasql/iasql-engine/some-pr-link-here) YYYY-MM-DD
+- [ ] Revoked/Superceded by: [RFC ###](./000 - RFC Template.md) YYYY-MM-DD
+
+## Author(s)
+
+- Joseph <joseph@iasql.com>
+
+## Summary
+
+Customers (even the very technically experienced ones) often have to look up how to construct an AWS IAM policy and then spend time validating the policy before inserting into IaSQL.
+This becomes more and more unmanageable as the policy grows.
+This is not a good experience, and it can be improved upon. We believe we can collect some inputs from IaSQL customers and offer then some form of automation when creating policies, so they can do it faster, more confidently and more conveniently.
+
+## Proposal
+
+We should add new `Policy` and `Statement` entities (perhaps in the `aws_iam` module). These entities would serve as 
+a way to construct IAM policies in a more user-friendly way. For example, the customer intends to create an S3 bucket with
+read-only permissions from all Principals within the same AWS account. Ideally, something like this:
+
+```json
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Resource": "aws:arn:123456789012:bucket-resource",
+      "Action": [
+        "s3:GetObjects"
+      ],
+      "Principal": {
+        "AWS": "arn:aws:iam::123456789012:root"
+      }
+    }
+  ]
+}
+```
+
+Instead of the customer having to look up and put together a JSON blob like this and then verifying and then inserting, we can also give them
+the option to do this using functions that would generate this blob.
+
+```sql
+-- User creates policy
+select * from iasql_create_policy(
+    'some_policy_name',
+    '2012-10-17' -- nullable, defaults to 2012-10-17
+    );
+
+-- User creates statement and attaches to the policy
+select * from iasql_create_policy_statement(
+        '<policy_id>',
+        'allow',
+        'aws:arn:123456789012:bucket-resource',
```

Comment:
Similar concern on the resources, however I am unsure if it is even possible to get a listing of all ARNs from the AWS SDK to use for validation purposes.

Reply:
Perhaps there is a way to check if an ARN string is a valid one, though?

Reply:
Do we really need to do this? From my observations, TF, CDK, CFN and others don't yet have this kind of validation, except at `apply` time.

If we absolutely do need to, then perhaps we use a custom Fn/graphile to make a call using the AWS SDK to validate the entire policy JSON (not just the ARN).
ref: https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/AccessAnalyzer.html#validatePolicy-property

We can also just do a quick regex check on properties like `Resource` ARNs, for simplicity.
A regex check will work for `Actions` too and we can also check them against an enum/set of enums.

Reply:
Pushing more details on this in a minute

Reply:
There are a couple of different concerns I want to address here:

1. On the comparison vs Terraform, CDK, etc. You are right that other deployment tools are not solving this issue, but that's part of why IaSQL can actually make sense. Better, earlier enforcement of what you're constructing in the database, enforced by foreign keys, column types/constraints, etc, should make `apply` less of a *gamble* than it is today. My concern/question here is: is this an unavoidable issue with AWS's design, or can we do better than the competition here?
2. This is not to say that I would *require* a solution here. "Perfect is the enemy of good" and all. If the time/resources to fix this are deemed too high by the RFC process itself, then we can note it as a "nice to have" in the future. If the RFC process says its impossible, then we can not worry about it at all. But if the RFC process says, "yeah, we can do that, and it'd add an extra hour or day of development work," then it seems reasonable to me to modify the RFC to include that as a requirement for the eventual implementation.